smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/UnionGenerator.java

```diff
@@ -203,5 +224,55 @@ private void writeVisitorFunction() {
         }
         writer.write("return visitor._(value.$$unknown[0], value.$$unknown[1]);");
         writer.dedent().write("}");
+        writer.write("");
+    }
+
+    private void writeFilterSensitiveLog() {
+        String objectParam = "obj";
+        writer.openBlock("export const filterSensitiveLog = ($L: $L): any => {", "}",
+            objectParam, symbol.getName(),
+            () -> {
+                for (MemberShape member : shape.getAllMembers().values()) {
+                    Shape memberTarget = model.expectShape(member.getTarget());
+                    String memberName = symbolProvider.toMemberName(member);
+
+                    if (member.getMemberTrait(model, SensitiveTrait.class).isPresent()) {
+                        // member is Sensitive, hide the value.
+                        writer.write("if (${1L}.${2L} !== undefined) return {[${2L}]: SENSITIVE_STRING};",
+                            objectParam, memberName);
+                    } else if (memberTarget instanceof SimpleShape) {
+                        writer.write("if (${1L}.${2L} !== undefined) return {${2L}: ${1L}.${2L}};",
+                            objectParam, memberName);
+                    } else {
+                        StructuredMemberWriter structuredMemberWriter = new StructuredMemberWriter(
+                            model, symbolProvider, shape.getAllMembers().values());
+                        String memberParam = String.format("%s.%s", objectParam, memberName);
+
+                        writer.write("if (${1L}.${2L} !== undefined) return {${2L}: ", objectParam, memberName);
+                        if (memberTarget instanceof StructureShape || memberTarget instanceof UnionShape) {
+                            structuredMemberWriter.writeStructureFilterSensitiveLog(writer, memberTarget, memberParam);
+                        } else if (memberTarget instanceof CollectionShape) {
+                            MemberShape collectionMember = ((CollectionShape) memberTarget).getMember();
```

Comment:
You can use the `.asCollectionShape().get()` pattern for these instead of casting.